avrdudes/avrdude
# Provide terminal config command to set fuses and lock bits

Never again fiddle with these pesky fuse calculators!

```
$ avrdude -c dryrun -p AVR32DA32 -t
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9533 (probably avr32da32)

avrdude> c
config period=t_off # 0
config window=t_off # 0
config sleep=bod_disabled # 0
config active=bod_disabled # 0
config sampfreq=sf_128hz # 0
config lvl=bod_1v9 # 0
config clksel=oschf # 0
config eesave=eex_erased # 0
config rstpincfg=gpio # 0
config crcsel=crc16 # 0
config crcsrc=nocrc # 0b11 = 3
config sut=sut_0ms # 0
config codesize=0 # 0
config bootsize=0 # 0
config key=nolock # 0x5cc5c55c

avrdude> c active= -vv
# Mask 0x0c of bodcfg: brownout detection in active/idle mode
config active=bod_disabled # 0 (brownout detection disabled)
# config active=bod_continuous # 1 = 4 >> 2 (brownout detection enabled in continuous mode)
# config active=bod_sampled # 0b10 = 2 = 0x08 >> 2 (brownout detection in sampled mode)
# config active=bod_continuous_wait # 0b11 = 3 = 0x0c >> 2 (brownout detection enabled in continuous mode with wake-up halted until BOD is ready)

avrdude> c sleep=3
avrdude warning: (config) assigning a reserved value (0x03) to sleep, check data sheet

avrdude> c -v sleep=
# Brownout detection in sleep mode
# config sleep=bod_disabled # 0
# config sleep=bod_continuous # 1
# config sleep=bod_sampled # 0b10 = 2
config sleep=0b11 # reserved = 3

avrdude> c sleep=bod_cont

avrdude> config -?
Syntax: config {<opts> | <property>[=<value>]}
Function: Show or change configuration properties of the part
Options:
    -f show value of fuse and lock bit memories as well
    -a output an initialisation script with all possible assignments
    -v increase verbosity, show explanations alongside output
    -h show this help message

Config alone shows all property names and current settings of the part's
hardware configuration in terms of symbolic mnemonics or values. Use

avrdude> config <property>

to show that of <property>. Wildcards or initial strings are permitted (but
not both), in which case all settings of matching properties are displayed.

avrdude> config <property>=

shows all possible values that <property> can take on with the currently
set one being the only that is not commented out. Assignments

avrdude> config <property>=<value>

modify the corresponding fuse or lock bits immediately but will normally only
take effect the next time the part is reset (see the data sheet). Value can
be a valid integer or one of the symbolic mnemonics, if known. Wildcards or
initial strings are permitted for the mnemonic, but an assignment only
happens if both the property and the name can be uniquely resolved.

It is quite possible, as is with direct writing to the underlying fuse and
lock bits, to brick a part, i.e., make it unresponsive to further programming
with the chosen programmer: here be dragons.
```

How to test?
 - Randomly select 20 parts and locate their data sheets
 - Play with `avrdude -c dryrun -p ... -t` for each of the parts and cross check with data sheet
 - Run the followg bash script
   ```
   #!/bin/bash
   mkdir -p /tmp/p
   for i in $(avrdude -p*/d | grep = | cut -f2 -d"'"); do
     echo "c -a" | avrdude -qqc dryrun -p$i -t 2> /tmp/p/$i.err | grep -v ^avrdude >/tmp/p/$i.config
     echo $i;
   done
   find /tmp/p/ -empty -delete
   ```
   Randomly check a few `/tmp/p/*.config` files
 - Now for the real test: buy a sack of parts (you will brick some)
 - Acquire a fire extinguisher (for all the High-Voltage reprogramming that you might need)
 - Test for real

Bugs: Only caters for 323 parts; does not play with `AT32UC3A0512` `AT89S51` `AT89S52` `ATxmega128A4` `ATxmega192A1` `ATxmega256A1` `ATxmega64A4` `AVR16EA28` `AVR16EA32` `AVR16EA48` `AVR32EA28` `AVR32EA32` `AVR32EA48` `AVR8EA28` `AVR8EA32` `LGT8F168P` `LGT8F328P` `LGT8F88P`
